# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmGenerator.cpp
# Commit: 4c22ce111c01
# Full Hash: 4c22ce111c012805515250b63074235b5578caa6
# Author: Ryan Hunt <rhunt@eqrion.net>
# Date: 2024-10-31 04:51:12
# Regressor Bug: 1913119
# File Overlap Count: 1
# Description:
#   Bug 1913119 - wasm: Re-use pool mechanism for lazy stubs and simplify logic. r=jseward
#   
#   Lazy func export stubs used some of the mechanism for allocating from a pool
#   of code segments, but not all. This caused some code duplication and layering
#   issues.
# ==============================================================================

diff -r 7593ec9fa478 -r 4c22ce111c01 js/src/wasm/WasmGenerator.cpp
--- a/js/src/wasm/WasmGenerator.cpp	Wed Oct 30 21:54:24 2024 +0000
+++ b/js/src/wasm/WasmGenerator.cpp	Wed Oct 30 21:54:24 2024 +0000
@@ -977,10 +977,9 @@
     // GC here as we may be running in OOL-code that is not ready for a GC.
     uint8_t* codeStart = nullptr;
     uint32_t codeLength = 0;
-    uint32_t metadataBias = 0;
-    codeBlock_->segment = CodeSegment::createFromMasmWithBumpAlloc(
-        *masm_, *linkData_, partialTieringCode_, /* allowLastDitchGC = */ false,
-        &codeStart, &codeLength, &metadataBias);
+    codeBlock_->segment = partialTieringCode_->createFuncCodeSegmentFromPool(
+        *masm_, *linkData_, /* allowLastDitchGC = */ false, &codeStart,
+        &codeLength);
     if (!codeBlock_->segment) {
       warnf("failed to allocate executable memory for module");
       return nullptr;
@@ -988,11 +987,11 @@
     codeBlock_->codeBase = codeStart;
     codeBlock_->codeLength = codeLength;
 
-    // In `codeBlock_`s metadata, we have a bunch of offsets which are
-    // relative to the start of the segment.  But we're placing the code at
-    // `metadataBias` forwards from the start of the segment, so we have to
-    // swizzle the metadata offsets accordingly.
-    codeBlock_->offsetMetadataBy(metadataBias);
+    // All metadata in code block is relative to the start of the code segment
+    // we were placed in, so we must adjust offsets for where we were
+    // allocated.
+    uint32_t codeBlockOffset = codeStart - codeBlock_->segment->base();
+    codeBlock_->offsetMetadataBy(codeBlockOffset);
   } else {
     // Create a new CodeSegment for the code and use that.
     codeBlock_->segment = CodeSegment::createFromMasm(